ivadomed/loader/loader.py

```diff
@@ -424,6 +424,83 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
 
         return dreturn
 
+    def read_file(self, filename):
+        """Read file according to file type.
+        Args:
+            filename (str): Subject filename.
+        """
+        if "nii" in self.extension:
+            # Returns 'nibabel.nifti1.Nifti1Image' object
+            # For 'nii' and 'nii.gz' extentions
+            return nib.load(filename)
+        elif self.extension == "png":
+            # Returns data from file as a 3D numpy array
+            # Behavior for grayscale png only, behavior TBD for RGB or RBGA png
+            return np.expand_dims(imageio.imread(filename, as_gray=True), axis=-1)
+        else:
+            raise RuntimeError('Input file type not supported')
+
+    def apply_canonical(self, data):
+        """Apply nibabel as_closest_canonical function to nifti data only.
+        Args:
+            data ('nibabel.nifti1.Nifti1Image' object or 'ndarray'):
+                for nifti or png file respectively.
+        """
+        if "nii" in self.extension:
+            # Returns 'nibabel.nifti1.Nifti1Image' object
+            return nib.as_closest_canonical(data)
+        if self.extension == "png":
+            # Returns data as is in numpy array
+            return data
+
+    def get_shape(self, data):
+        """Returns data shape according to file type.
+        Args:
+            data ('nibabel.nifti1.Nifti1Image' object or 'ndarray'):
+                for nifti or png file respectively.
+        Returns:
+            ndarray: Data shape.
+        """
+        if "nii" in self.extension:
+            return data.header.get_data_shape()
+        if self.extension == "png":
+            return data.shape
+
+    def get_voxel_size(self, data):
+        """Returns voxel sizes in mm according to file type.
+        Args:
+            data ('nibabel.nifti1.Nifti1Image' object or 'ndarray'):
+                for nifti and png file respectively.
+        Returns:
+            tuple: Voxel size in mm
+        """
+        if "nii" in self.extension:
+            # Read zooms metadata from nifti file header
+            return data.header.get_zooms()
+        if self.extension == "png":
+            # Voxel size for PNG is extracted from PixelSize metadata (from BIDS JSON sidecar)
+            # PixelSize definition in example dataset is a scalar in micrometers (BIDS BEP031 v 0.0.2)
+            # PixelSize definition will change, 2D (XY) or 3D (XYZ) array in micrometers (BIDS BEP031 v 0.0.3)
+            # Behavior will have to be ajusted here to follow BEP development
+            pixel_size_in_mm = self.metadata[0]['PixelSize'] / 1000
+            return (pixel_size_in_mm, pixel_size_in_mm, 0)
```

Comment:
This worries me a little, because we rely on `PixelSize` to be defined the JSON sidecar. Is there a possibility that this info could ever be missing? 

If so, then perhaps we could provide a more informative error message.

Reply:
Side question: Does the BIDS JSON sidecar have a field for PixelSize units (e.g. `um` for micrometers)? That would make our life a lot easier, hehe.

Reply:
As it is defined now, the `PixelSize` in BIDS JSON sidecar is always in `micrometers`, so I just converted to `mm`. However,  microscopy is not integrated in the BIDS specification yet. We are working on an BIDS extension proposal (BEP) in parallel, so microscopy data/metadata may change.

I updated the function to add an error message and support PixelSize as a float or as a list.

Reply:
As defined in BIDS right now 'PixelSize' is always in `micrometer `so I just converted to `mm`. However, microscopy is not integrated in the BIDS specification yet. We are working in parallel on a BIDS extension proposal (BEP) for it and microscopy data/metadata may change.

I added an error message if the field is not present and support for `PixelSize` as a float or as a list.

Reply:
Thank you for the clarification. :slightly_smiling_face: